File: backend/app/schemas/document.py
"""Home Digital Twin document schemas — mirror frontend/lib/types/document.ts and
the `docs` table (backend/db/schema.sql). One shape for all five categories
(invoice, warranty, photo, manual + service records handled separately) rather
than a class per category, since the DB uses a single table + metadata JSONB.
"""

from typing import Literal

from app.schemas.user import CamelModel

DocumentCategory = Literal["invoice", "warranty", "photo", "manual"]


class DocumentOut(CamelModel):
    id: str
    category: DocumentCategory
    title: str
    file_url: str | None = None
    file_type: str
    uploaded_at: str
    tags: list[str] = []
    linked_appliance: str | None = None
    notes: str | None = None
    # Invoice-specific (also usable by other categories where relevant)
    vendor: str | None = None
    amount: int | None = None
    purchase_date: str | None = None
    order_number: str | None = None
    # Warranty-specific
    brand: str | None = None
    expires_at: str | None = None


class DocumentCreate(CamelModel):
    """Request body for POST /documents. No real file upload yet (see
    Phase 1 of docs/../BestBuild-Plans/01-homeowner-digital-twin.md — object
    storage isn't wired up); fileUrl is a client-supplied placeholder for now.
    """

    category: DocumentCategory
    title: str
    file_url: str | None = None
    file_type: str = "pdf"
    tags: list[str] = []
    linked_appliance: str | None = None
    notes: str | None = None
    vendor: str | None = None
    amount: int | None = None
    purchase_date: str | None = None
    order_number: str | None = None
    brand: str | None = None
    expires_at: str | None = None

File: frontend/lib/homeowner/types/document.ts
/** Home Digital Twin types — mirror backend/app/homeowner/schemas/document.py
 * and the `docs` table. One flat shape per concept (not a class per category)
 * since the DB stores category-specific fields in a single metadata JSONB
 * column rather than a table per category.
 */

export type DocumentCategory = "invoice" | "warranty" | "photo" | "manual";

export interface HomeDocument {
  id: string;
  category: DocumentCategory;
  title: string;
  fileUrl: string | null;
  fileType: string;
  uploadedAt: string;
  tags: string[];
  linkedAppliance: string | null;
  notes: string | null;
  // Invoice-specific (also usable elsewhere where relevant)
  vendor?: string | null;
  amount?: number | null;
  purchaseDate?: string | null;
  orderNumber?: string | null;
  // Warranty-specific
  brand?: string | null;
  expiresAt?: string | null;
}

/** Fields the "Add document" form can submit. No real file upload yet —
 * object storage isn't wired up. */
export interface HomeDocumentCreate {
  category: DocumentCategory;
  title: string;
  fileType?: string;
  tags?: string[];
  linkedAppliance?: string | null;
  notes?: string | null;
  vendor?: string | null;
  amount?: number | null;
  purchaseDate?: string | null;
  orderNumber?: string | null;
  brand?: string | null;
  expiresAt?: string | null;
}
